apm-agent-plugins/apm-log-shader-plugin/apm-log4j2-plugin/src/main/java/co/elastic/apm/agent/log4j2/Log4j2AppenderGetLayoutAdvice.java

```diff
@@ -0,0 +1,61 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.log4j2;
+
+import co.elastic.apm.agent.sdk.advice.AssignTo;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.implementation.bytecode.assign.Assigner;
+import org.apache.logging.log4j.core.Appender;
+import org.apache.logging.log4j.core.Layout;
+
+import javax.annotation.Nullable;
+import java.io.Serializable;
+
+/**
+ * The Log4j2 {@link Appender} does not expose a {@code setLayout()} API that allows us to override logging events.
+ * However, it exposes an {@link Appender#getLayout()} API <b>that is always used when events are logged</b>.
+ * Therefore, by instrumenting this method and replacing the returned {@link Layout}, we can implement the
+ * {@link co.elastic.apm.agent.logging.LogEcsReformatting#OVERRIDE OVERRIDE} use case.
+ */
+public class Log4j2AppenderGetLayoutAdvice {
```

Comment:
I think we should agree on where to put advices. Either as static inner classes or as stand-alone classes.

I'd prefer the static inner class approach a lot. It makes instrumentations more cohesive as it's easier to grasp where an advice is applied to.

Reply:
I agree it's much more readable this way, the thing is that it's much easier to get it wrong and make the instrumentation independent on library types. I fixed quite a few of those lately, and I think a main reason for that is allowing something in the `import` section, that is not allowed somewhere in the class.

Reply:
Are you thinking about automatically validating the imports? Or are you saying that it's just easier for the reviewer to verify that there are no invalid imports?

I just had an idea how we could verify that the instrumentation doesn't use invalid classes. In `AbstractInstrumentationTest`, we could create an isolated child class loader (parent=bootstrap) that loads each instrumentation class and calls the `get*Matcher` methods. That would result in linkage errors if library types are referenced.

Reply:
> Are you thinking about automatically validating the imports? Or are you saying that it's just easier for the reviewer to verify that there are no invalid imports?

Auto-validating the imports means looking at source, so probably not. My experience shows that it's easier for the author to add a reference to something that is already at the imports in the IDE, and easier for the reviewer to miss.

> I just had an idea how we could verify that the instrumentation doesn't use invalid classes. In AbstractInstrumentationTest, we could create an isolated child class loader (parent=bootstrap) that loads each instrumentation class and calls the get*Matcher methods. That would result in linkage errors if library types are referenced.

Good idea! You mean a separate URL CL that knows the agent jar and is the child of the boot CL (since it's a unit test, we cannot rely on the boot CL to load agent classes)? 
The only thing is that we will still rely only on tested JVMs, and may be exposed to differences in linkage in non tested ones, but should capture such issues very quickly in most cases.